pkg/vmprovider/providers/vsphere/vmprovider_vm_utils.go

```diff
@@ -89,117 +94,159 @@ func GetVMImageStatusAndContentLibraryUUID(
-			break
+		csBindingList := &vmopv1alpha1.ContentSourceBindingList{}
+		if err := k8sClient.List(vmCtx, csBindingList, ctrlclient.InNamespace(vmCtx.VM.Namespace)); err != nil {
+			msg := fmt.Sprintf("Failed to list ContentSourceBindings in namespace: %s", vmCtx.VM.Namespace)
+			conditions.MarkFalse(vmCtx.VM,
+				vmopv1alpha1.VirtualMachinePrereqReadyCondition,
+				vmopv1alpha1.ContentSourceBindingNotFoundReason,
+				vmopv1alpha1.ConditionSeverityError,
+				msg)
+			vmCtx.Logger.Error(err, msg)
+			return nil, "", errors.Wrap(err, msg)
 		}
-	}
-	if contentSourceName == "" {
-		msg := fmt.Sprintf("ContentLibraryProvider %s does not have a ContentSource OwnerReference", clProvider.Name)
-		conditions.MarkFalse(vmCtx.VM,
-			vmopv1alpha1.VirtualMachinePrereqReadyCondition,
-			vmopv1alpha1.ContentSourceBindingNotFoundReason,
-			vmopv1alpha1.ConditionSeverityError,
-			msg)
-		return nil, "", errors.New(msg)
+
+		// Filter the bindings for the specified ContentSource.
+		matchingContentSourceBinding := false
+		for _, csBinding := range csBindingList.Items {
+			if csBinding.ContentSourceRef.Kind == "ContentSource" && csBinding.ContentSourceRef.Name == contentSourceName {
+				matchingContentSourceBinding = true
+				break
+			}
+		}
+
+		if !matchingContentSourceBinding {
+			msg := fmt.Sprintf("Namespace %s does not have access to ContentSource %s for VirtualMachineImage %s",
+				vmCtx.VM.Namespace, clUUID, vmImage.Name)
+			conditions.MarkFalse(vmCtx.VM,
+				vmopv1alpha1.VirtualMachinePrereqReadyCondition,
+				vmopv1alpha1.ContentSourceBindingNotFoundReason,
+				vmopv1alpha1.ConditionSeverityError,
+				msg)
+			return nil, "", errors.New(msg)
+		}
+
+		vmImageStatus = &vmImage.Status
+		vmImageSpec = &vmImage.Spec
 	}
 
-	csBindingList := &vmopv1alpha1.ContentSourceBindingList{}
-	if err := k8sClient.List(vmCtx, csBindingList, ctrlclient.InNamespace(vmCtx.VM.Namespace)); err != nil {
-		msg := fmt.Sprintf("Failed to list ContentSourceBindings in namespace: %s", vmCtx.VM.Namespace)
-		conditions.MarkFalse(vmCtx.VM,
-			vmopv1alpha1.VirtualMachinePrereqReadyCondition,
-			vmopv1alpha1.ContentSourceBindingNotFoundReason,
-			vmopv1alpha1.ConditionSeverityError,
-			msg)
-		vmCtx.Logger.Error(err, msg)
-		return nil, "", errors.Wrap(err, msg)
+	if err = checkVMImageHWVersion(vmCtx, vmImageSpec); err != nil {
+		return nil, "", err
 	}
 
-	// Filter the bindings for the specified ContentSource.
-	matchingContentSourceBinding := false
-	for _, csBinding := range csBindingList.Items {
-		if csBinding.ContentSourceRef.Kind == "ContentSource" && csBinding.ContentSourceRef.Name == contentSourceName {
-			matchingContentSourceBinding = true
-			break
+	return vmImageStatus, clUUID, nil
+}
+
+// checkVMImageHWVersion checks if the VirtualMachineImage's HW version is at least the minimum supported virtual hw version.
+func checkVMImageHWVersion(vmCtx context.VirtualMachineContext, imageSpec *vmopv1alpha1.VirtualMachineImageSpec) error {
```

Comment:
Instead of having this as a nested function, I suggest making it public and calling it directly from vm_provider after getting the imageSpec. This will help make it easier for someone to locate the code responsible for setting the VM spec condition.